src/goto-harness/recursive_initialization.cpp

```diff
@@ -136,5 +151,180 @@ void recursive_initializationt::initialize_nondet(
   code_blockt &body)
 {
-  body.add(code_assignt{lhs, side_effect_expr_nondett{lhs.type()}});
+  if(
+    lhs.id() != ID_symbol ||
+    !is_array_size_parameter(to_symbol_expr(lhs).get_identifier()))
+  {
+    body.add(code_assignt{lhs, side_effect_expr_nondett{lhs.type()}});
+  }
+}
+
+void recursive_initializationt::initialize_array(
+  const exprt &array,
+  std::size_t depth,
+  const recursion_sett &known_tags,
+  code_blockt &body)
+{
+  PRECONDITION(array.type().id() == ID_array);
+  const auto &array_type = to_array_type(array.type());
+  const auto array_size =
+    numeric_cast_v<std::size_t>(to_constant_expr(array_type.size()));
+  for(std::size_t index = 0; index < array_size; index++)
+  {
+    initialize(
+      index_exprt(array, from_integer(index, size_type())),
+      depth,
+      known_tags,
+      body);
+  }
+}
+
+bool recursive_initializationt::should_be_treated_as_array(
+  const irep_idt &array_name) const
+{
+  return initialization_config.pointers_to_treat_as_arrays.find(array_name) !=
+         initialization_config.pointers_to_treat_as_arrays.end();
+}
+
+bool recursive_initializationt::is_array_size_parameter(
+  const irep_idt &cmdline_arg) const
+{
+  return initialization_config.variables_that_hold_array_sizes.find(
+           cmdline_arg) !=
+         initialization_config.variables_that_hold_array_sizes.end();
+}
+
+optionalt<irep_idt> recursive_initializationt::get_associated_size_variable(
+  const irep_idt &array_name) const
+{
+  return optional_lookup(
+    initialization_config.array_name_to_associated_array_size_variable,
+    array_name);
+}
+
+void recursive_initializationt::initialize_dynamic_array(
+  const exprt &pointer,
+  std::size_t depth,
+  const recursion_sett &known_tags,
+  code_blockt &body)
+{
+  PRECONDITION(pointer.type().id() == ID_pointer);
+
+  // size_t number_of_arrays = max_array_size - min_array_size;
+  // T *arrays[number_of_arrays];
+  //
+  // T array1[min_array_size];
+  // initialize(array1);
+  // arrays[0] = &array1[0];
+  //
+  // T array2[min_array_size+1];
+  // initialize(array2);
+  // arrays[1] = &array2[0];
+  //
+  // ... and so on ...
+  //
+  // T arrayN[max_array_size];
+  // initialize(arrayN);
+  // arrays[number_of_arrays-1] = &arrayN[0];
+  //
+  // size_t array_index;
+  // assume(0 <= array_index < number_of_arrays);
+  // actual_size = array_index + min_array_size;
+  // array_pointer = arrays[array_index];
+
+  const auto &pointer_type = to_pointer_type(pointer.type());
+  allocate_objectst allocate_objects{initialization_config.mode,
+                                     pointer_type.source_location(),
+                                     "initializer",
+                                     goto_model.symbol_table};
+  auto min_array_size = initialization_config.min_dynamic_array_size;
+  auto max_array_size = initialization_config.max_dynamic_array_size;
+  PRECONDITION(max_array_size >= min_array_size);
+  std::vector<symbol_exprt> array_variables;
```

Comment:
💡 I think you should group creation of these locals into immediately-invoked lambda expressions (at the very least). As it is, this function is very difficult to follow, it's very long, everything is mutable and potentially touched.

Reply:
There are definitely a couple places where variables that are meant to be `const` aren't marked as such and should be fixed. I'll take a look at how the function would look with immediately invoked lambdas to see if that'd actually improve the readability.